chunky/src/java/se/llbit/math/NodeBasedOctree.java

```diff
@@ -151,334 +184,52 @@ public void store(DataOutputStream out) throws IOException {
-        }
-      }
-
-      t = ((lz << level) - ray.o.z) / ray.d.z;
-      if (t < tNear && t > Ray.EPSILON) {
-        tNear = t;
-        nz = 1;
-        nx = ny = 0;
-      } else {
-        t = (((lz + 1) << level) - ray.o.z) / ray.d.z;
-        if (t < tNear && t > Ray.EPSILON) {
-          tNear = t;
-          nz = -1;
-          nx = ny = 0;
-        }
-      }
+  public long nodeCount() {
+    return countNodes(root);
+  }
 
-      ray.o.scaleAdd(tNear, ray.d);
-      ray.n.set(nx, ny, nz);
-      ray.distance += tNear;
+  private long countNodes(Octree.Node node) {
+    if(node.type == BRANCH_NODE) {
+      long total = 1;
+      for(int i = 0; i < 8; ++i)
+        total += countNodes(node.children[i]);
+      return total;
+    } else {
+      return 1;
     }
   }
 
-  /**
-   *  {@inheritDoc}
-   */
-  @Override
-  public boolean exitWater(Scene scene, Ray ray, BlockPalette palette) {
-    if (!isInside(ray.o) && !enterOctree(ray))
-      return false;
-
-    // Marching is done in a top-down fashion: at each step, the octree is descended from the root to find the leaf
-    // node the ray is in. Terminating the march is then decided based on the block type in that leaf node. Finally the
-    // ray is advanced to the boundary of the current leaf node and the next, ready for the next iteration.
-    while (true) {
-      // Add small offset past the intersection to avoid
-      // recursion to the same octree node!
-      int x = (int) QuickMath.floor(ray.o.x + ray.d.x * Ray.OFFSET);
-      int y = (int) QuickMath.floor(ray.o.y + ray.d.y * Ray.OFFSET);
-      int z = (int) QuickMath.floor(ray.o.z + ray.d.z * Ray.OFFSET);
-
-      int lx = x >>> depth;
-      int ly = y >>> depth;
-      int lz = z >>> depth;
-
-      if (lx != 0 || ly != 0 || lz != 0)
-        return false; // outside of octree!
-
-      // Descend the tree to find the current leaf node
-      Octree.Node node = root;
-      int level = depth;
-      while (node.type == BRANCH_NODE) {
-        level -= 1;
-        lx = x >>> level;
-        ly = y >>> level;
-        lz = z >>> level;
-        node = node.children[((lx & 1) << 2) | ((ly & 1) << 1) | (lz & 1)];
-      }
-
-      // Test intersection
-      Block currentBlock = palette.get(node.type);
-      Material prevBlock = ray.getCurrentMaterial();
-
-      ray.setPrevMaterial(prevBlock, ray.getCurrentData());
-      ray.setCurrentMaterial(currentBlock, node.type);
-
-      if (!currentBlock.isWater()) {
-        if (currentBlock.localIntersect) {
-          if (!currentBlock.intersect(ray, scene)) {
-            ray.setCurrentMaterial(Air.INSTANCE, 0);
-          }
-          return true;
-        } else if (currentBlock != Air.INSTANCE) {
-          TexturedBlockModel.getIntersectionColor(ray);
-          return true;
-        } else {
-          return true;
-        }
+  static public void initImplementation() {
+    Octree.addImplementationFactory("NODE", new Octree.ImplementationFactory() {
```

Comment:
Since `"NODE"` is used all over the place, maybe we should move it into a constant (`NodeBasedOctree.NAME`)? Not sure about it, though.